### PR TITLE
fix: double quotes will be trimmed from the search token

### PR DESCRIPTION
**Pull request by jvillafanez**
## Description
Search terms from the graph API can be enclosed with double quotes. The double quotes are intended to prevent the interpretation of some tokens from the parser. However, the parser doesn't remove the double quotes.
This PR will remove those double quotes.

As a practical example, searching by email (such as `marie@example.org`) can give problems because the `@` char is interpreted as a different token (likely as a parameter alias, although I'm not fully sure). With this PR, you can search it by enclosing the email in double quotes, like `$search="marie@example.org"` (note that url-encoding might be needed)

## Related Issue
https://github.com/owncloud/ocis/issues/7990

## Motivation and Context
Allow searching users by email

## How Has This Been Tested?
Manually tested via API, running something like `curl -vk -ueinstein:relativity https://localhost:9200/graph/v1.0/users?%24search=%22marie%40example.org%22`

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:

- [x] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link>

### CODE REVIEW

File: services/graph/pkg/identity/odata.go [hunk 1]
```diff
@@ -37,5 +41,9 @@ func GetSearchValues(req *godata.GoDataQuery) (string, error) {
 		return "", godata.NotImplementedError("complex search queries are not supported")
 	}
 
-	return req.Search.Tree.Token.Value, nil
+	searchValue := req.Search.Tree.Token.Value
+	if strings.HasPrefix(searchValue, "\"") && strings.HasSuffix(searchValue, "\"") {
+		searchValue = strings.Trim(searchValue, "\"")
+	}
```
Comment (rhafer) [inline services/graph/pkg/identity/odata.go:47]: Nitpick: I guess in this case you can just call `strings.Trim` unconditionally. There is no need for the `HasPrefix` and `HasSuffix` calls. Or am I missing something?

The OData parser will already choke on strings that only contain a single double quote.

**Review by rhafer — APPROVED** [review]
lgtm. Just a minor nitpick. Feel free to ignore.

**Review by rhafer — CHANGES_REQUESTED** [review]
Hm, actually I think I overlooked something. We still have minimum search length check in https://github.com/owncloud/ocis/blob/master/services/graph/pkg/service/v0/users.go#L226.
Where check of the length of the possibly quoted string. I think we need to adjust that. Otherwise it would be simple to circumvent that limit (or at least reduce it by 2) by just adding quotes.

File: services/graph/pkg/identity/odata.go (hunk 1 above)
Comment (jvillafanez) [reply services/graph/pkg/identity/odata.go:47]: > The OData parser will already choke on strings that only contain a single double quote.

I thought it was still possible. I'll remove the condition.

Comment (mmattel) [reply services/graph/pkg/identity/odata.go:47]: How are we dealing the same with single quotes?

Comment (rhafer) [reply services/graph/pkg/identity/odata.go]: I think single quotes have a completely different meaning here and are handled by the OData parser.  In this case, as far as I understand the odata specs, using single quotes here won't even work. (our parsers doesn't allow it here at least https://github.com/CiscoM31/godata)

**Review by rhafer — APPROVED** [review]

File: services/graph/pkg/identity/odata.go (hunk 1 above)
Comment (mmattel) [reply services/graph/pkg/identity/odata.go]: does a _user_ (the one who is on the keyboard...) know this difference...?

Comment (rhafer) [reply services/graph/pkg/identity/odata.go]: The user shouldn't need to worry about quoting at all.

The UI/client which is constructing the query should know the difference, yes.

**Comment by jvillafanez** [comment]
Refactored the code a bit. The tests should pass now.
I'll manually retest the code just to make sure it works with the changes.

**Comment by jvillafanez** [comment]
> I'll manually retest the code just to make sure it works with the changes.

Checked. Looks good now.

**Review by rhafer — APPROVED** [review]

**Comment by micbar** [comment]
I would love it so see a unit test for that 😍

**Comment by rhafer** [comment]
> I would love it to see a unit test for that 😍

They are part of #8050 now.

**Pull request merged**